crates/core/src/nats/client.rs

```diff
@@ -0,0 +1,223 @@
+use std::{sync::Arc, time::Duration};
+
+use bytes::Bytes;
+use tracing::info;
+
+use super::{
+    types::{
+        JetStreamConfig,
+        JetStreamContext,
+        NatsConsumer,
+        NatsStream,
+        PullConsumerConfig,
+    },
+    NatsError,
+    Subject,
+};
+use crate::types::BoxedResult;
+
+#[derive(Debug, Clone)]
+pub struct NatsClient {
+    pub url: String,
+    pub conn_id: String,
+    pub conn: Arc<async_nats::Client>,
+    pub jetstream: Arc<JetStreamContext>,
+}
+
+impl NatsClient {
+    pub async fn connect(
+        url: &str,
+        conn_id: &str,
+        nkey: Option<String>,
+    ) -> Result<Self, NatsError> {
+        let conn_id = conn_id.to_string();
+        let conn = Self::create_conn(url, nkey).await?;
+        let context =
+            Arc::new(async_nats::jetstream::new(conn.as_ref().clone()));
+
+        info!("Connected to NATS server at {}", url);
+        Ok(Self {
+            url: url.to_string(),
+            conn_id,
+            conn,
+            jetstream: context,
+        })
+    }
+
+    async fn create_conn(
+        url: &str,
+        nkey: Option<String>,
```

Comment:
Why not:

```suggestion
        nkey: &str,
```

Reply:
If you look below, it's returning an error when nkey is none

Reply:
Yes, that's the point. We don't need to worry about that error if we type-enforce that an NKEY must always be provided

Reply:
Got your point, makes sense